### PR TITLE
Audit Item M02

### LINKED ISSUE #21: Resolve audit item M02

### PR DESCRIPTION
**Pull request by apbendi**
WIP: Currently implements 1/2 of M02, specifically related to claims.

TODO: The other half of M02, related to grants.

@CantelopePeel @wildmolasses @mds1 please take a careful look at what I've got so far, see if it makes sense to you and meets the needs both of the audit and the broader application in general.

### CODE REVIEW

File: contracts/Org.sol
Comment (CantelopePeel) [inline contracts/Org.sol]: Add a comment that states the mapping is from UUIDs to Claims

File: contracts/Org.sol
Comment (CantelopePeel) [inline contracts/Org.sol:39]: If we have active claim this is not needed right?

File: contracts/Org.sol
Comment (CantelopePeel) [inline contracts/Org.sol:72]: does this need to be marked memory?

File: contracts/Org.sol
Comment (CantelopePeel) [inline contracts/Org.sol]: require stmt to check for claimId existence.

File: contracts/Org.sol
Comment (CantelopePeel) [inline contracts/Org.sol:79]: Note: make sure orgAdminWalletAddress is not zero when merging in M01

File: contracts/Org.sol
Comment (CantelopePeel) [inline contracts/Org.sol]: Why is this public and not external now?

File: test/Org.test.js
Comment (CantelopePeel) [inline test/Org.test.js:152]: Style for this looks different than it does below

File: test/Org.test.js
Comment (CantelopePeel) [inline test/Org.test.js:170]: See previous

**Review by CantelopePeel — CHANGES_REQUESTED** [review]

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol:39]: The taxId is for the Organization, and is set when it's created/deployed. The claim does not have an EIN.

Comment (CantelopePeel) [reply contracts/Org.sol:39]: Oh duh yup

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol]: good idea, done!

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol:72]: Nope, not needed/allowed for the `address` type.

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol:79]: It seemed reasonable to me to ignore this check on claim submission, as long as we don't allow such a malformed claim to be approved or cashed out. It would be easy to add though, could go either way honestly.

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol]: What do you want to check, here? Just that it's not the empty string? I don't think that actually buys us much in terms of preventing harm.

File: contracts/Org.sol
Comment (CantelopePeel) [reply contracts/Org.sol:79]: This is resolved in M02 I just want to ensure it stays resolved... More a reminder to me that it doesn't slip through when I merge it.

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol]: Good catch! Missed this on the rebase. Fixed.

File: test/Org.test.js
Comment (apbendi) [reply test/Org.test.js:152]: Fixed!

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol:79]: Not sure what you mean? This is M02.

Comment (apbendi) [reply contracts/Org.sol:79]: Ahh I see you probably meant M01? All good 👍

File: contracts/Org.sol
Comment (CantelopePeel) [reply contracts/Org.sol]: Add lines like:
```
  Claim storage claim = pendingClaims[claimId];
   require(claim.desiredWallet != address(0), "Org: Cannot reject a claim with no wallet");
```
I have a hunch the auditors will see this lacking a check and push it back to us.

**Review by CantelopePeel — CHANGES_REQUESTED** [review]

File: contracts/Org.sol
Comment (mds1) [inline contracts/Org.sol:40]: Could reduce gas costs in a lot of places by instead using a mapping from `bytes32 => Claim` where `bytes32` is the hash of the UUID

File: test/Org.test.js
Comment (mds1) [inline test/Org.test.js]: 123 is not a valid EIN, so this will fail when merged with #74. Using 999999999 or any other 9 digit number should work (there's a few places to change this)

**Review by mds1 — APPROVED** [review]
Looks good! A few small comments below but nothing major.

For `grants`, will there be a UUID generated for those as well? That way we can use a similar architecture of `mapping(string => Grant) public grants; `

File: contracts/Org.sol
Comment (wildmolasses) [inline contracts/Org.sol]: Questioning the utility of `rejectClaim`. Why not just `approveClaim` the good one? (nitpicking!)

File: contracts/Org.sol
Comment (wildmolasses) [inline contracts/Org.sol:104]: Is there a case where overwriting an already approved claim (`activeClaim`) should be prevented? Just thought I'd ask.

Comment (wildmolasses) [reply contracts/Org.sol:104]: Aside, would it make any sense to clear the pendingClaims entirely after a claim is approved?

File: contracts/Org.sol
Comment (rheeger) [reply contracts/Org.sol:40]: +1 good shout

File: contracts/Org.sol
Comment (rheeger) [reply contracts/Org.sol:104]: dont think we should clear pending claims after an approval. It's possible we'll approve the wrong one and wouldn't want to make an org resubmit the claim.

File: contracts/Org.sol
Comment (rheeger) [reply contracts/Org.sol]: +1 I agree here. Will reduce SLOC and confusion. Only think i can think of is if someone spams a org contract and drops a ton of claims in, would that impact contract interaction GAS prices/performance? Would want a method to clear those out if so.

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol:40]: If we did this, would we want to pass the hash in to the `requestClaim`, `acceptClaim`, and `rejectClaim` method, or the string itself?

If the latter, does calculating the hash and storing it still work out to less gas than storing a UUID sized string? (I'm guessing yes, but would want to confirm).

If the former, we won't be able to announce the string version of the UUID in the emitted event. This means, to know the UUID of existing claims, you'd have to get them from somewhere else, namely the database, or the endaoment API if you're a third party. Perhaps that's fine, just want to point out the tradeoff!

File: test/Org.test.js
Comment (apbendi) [reply test/Org.test.js]: Should be easy enough to do a find/replace. Should probably pull it into a constant variable at the top of the test anyway. @mds1 mind working that into your updates?

Comment (mds1) [reply test/Org.test.js]: Yea there's a constant called `ein` in #74 set to 999999999. Can probably do the same here, or just set it to a valid number and we can update during the rebase/merge

File: contracts/Org.sol
Comment (mds1) [reply contracts/Org.sol:40]: I was thinking we'd pass the hash in directly, but good point about events. We could just store the hash and UUID in the DB like you suggested, but I'm not totally sure who will be needing it, so will defer to @rheeger on that

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol]: Two thoughts:

1) If someone is observing the event logs to see the progression of claims, I thought it'd be useful to see a claim explicitly rejected. Maybe not needed, though.
2) If we don't have a method like this, it's not possible to clear the storage space. It doesn't cost US any gas, but it does mean the data will live in Ethereum's state bloat for all eternity.

Perhaps this is a bit of smell that the architecture is still off— maybe we want to disallow the submission of multiple claims altogether? Have a slot for a pendingClaim, it either get's accepted (clear pending, store in active) or rejected (just clear pending). Thoughts @rheeger / @CantelopePeel ?

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol:104]: Per comment above, if we don't clear the claim, it lives in Ethereum's state forever, which would make us bad blockchain citizens. From a selfish perspective here, deleting the state from `pendingClaims` would reduce the gas cost to approve.

File: contracts/Org.sol
Comment (apbendi) [reply contracts/Org.sol]: It's one thing to check if an address is 0. For claimId, though, we're expecting it to be a UUID string. Validating a properly formatted UUID string in Solidity would be painful, expensive, and (IMO) not add much value.

Comment (mds1) [reply contracts/Org.sol]: Ben has a good point that that we don't really even need an array, do we? 

> Have a slot for a pendingClaim, it either get's accepted (clear pending, store in active) or rejected (just clear pending)"

This sounds like a better approach to me.

The attack here is that someone runs a bot to continually fill the one pending claim with a junk claim, so to resolve this you can have an admin-only function like `rejectClaimAndApproveNewClaim` which additionally takes the same inputs as `claimRequest`. That way in one transaction you reject the existing claim and automatically approve the new one

File: contracts/Org.sol
Comment (CantelopePeel) [reply contracts/Org.sol:40]: A few questions: How much of a savings are we talking about? Won't a hash reduce the integrity (however marginally) of the UUID? I wonder cause users will want to see the uuid if they inspect the blockchain. @apbendi We can also calc hash off chain.

Comment (mds1) [reply contracts/Org.sol:40]: A quick storage test shows 46080 gas to save a string to a mapping, and 44599 gas to save the keccak256 hash of that string to a mapping. So it seems because the string is relatively short it doesn't make a big difference, so might be easiest to leave this as a string. Contract is below if anyone wants to reproduce:

```solidity
pragma solidity ^0.6.7;

contract StringTest {
     mapping (string => bool) public stringCheck;
     mapping (bytes32 => bool) public bytesCheck;
     
    function saveString(string memory _var) public {
        stringCheck[_var] = true;
    }

     function saveBytes(bytes32 _var) public {
        bytesCheck[_var] = true;
    }
}
```

File: contracts/Org.sol
Comment (CantelopePeel) [reply contracts/Org.sol]: Is there any other way to mitigate this potential vulnerability? I am not particularly satisfied with the `rejectClaimAndApproveNewClaim` idea, but if it is the only satisfactory way to mitigate this I will settle.

Comment (CantelopePeel) [reply contracts/Org.sol]: No I just want a check that the address is not 0. I agree it is not necessary to look at the UUID, that is our deal to ensure it is the correct value.

Comment (mds1) [reply contracts/Org.sol]: Hmm, no other good ones that I can think of currently. I personally think `rejectClaimAndApproveNewClaim` is a pretty clean solution, because you'll be verifying the claim is correct off-chain anyway, so in this event it should be simple to submit the info yourself instead of having the Org do it.

Curious to hear what you don't like about!

Comment (rheeger) [reply contracts/Org.sol]: Just want to make sure I understand how `rejectClaimAnApproveNewClaim` would work... 

Under this structure, only one `pendingClaim` could be submitted at a time until a decision was made about that pending claim. However, under the attack described by @mds1, someone could file a set of bogus claims and keep filing them after review to prevent anyone else from submitting a valid claim. 

`rejectClaimAndApproveNewClaim` should require either the `ADMIN` or `REVIEWER` roles to reject the given bogus `pendingClaim` while also providing the details of a valid claim to be atomically approved inside the method.

Is that accurate @mds1 ?  

If so, I think this introduces added complexity that is maybe unnecessary. It seems to me that we should go back to being able to hold multiple `pendingClaims` at a time. I know this would mean holding bloated contract state, but it inherently thwarts the described attack. 

Thoughts?

Comment (apbendi) [reply contracts/Org.sol]: Matt's point about the spam claims blocking legitimate ones is a good one. Robbie's point about the atomic update adding complexity is also a good one. Thinking through it, my preference would be to leave the system as is, including the `rejectClaim` method. If you choose not to use it, that's fine, but I don't think it having it has any downsides.

File: contracts/Fund.sol
Comment (wildmolasses) [inline contracts/Fund.sol:153]: Is `updateGrant` necessary? Took me 3 minutes to write; if yes, will write tests; if no, will delete.

File: contracts/Fund.sol
Comment (wildmolasses) [inline contracts/Fund.sol:133]: here I'm not allowing grants to be overwritten. Is this desirable? If we allow grants to be overwritten, then we don't need a separate `updateGrant` function. Less explicit though.

Then again, perhaps grants will never be updated and I'm overthinking it.  Hey, just going along with the prompt from M02!

File: contracts/OrgFactory.sol
Comment (wildmolasses) [inline contracts/OrgFactory.sol]: Do we care about disallowing orgs? Then, same thing here, should we prevent overwrites on the createOrg and write an explicit allowOrg function?

File: contracts/Fund.sol
Comment (wildmolasses) [inline contracts/Fund.sol]: Added this grantId array so we could still use `getGrantsCount` and also provide grant count in `getSummary`. Let me know if the count is unnecessary and i will remove the array, the `getGrantsCount` method, and the grant count in `getSummary`.

File: contracts/OrgFactory.sol
Comment (rheeger) [reply contracts/OrgFactory.sol]: This is a great point and should be something we keep. It is definitely possible that an org will fall out of good standing and need to be disallowed. The org could also dissolve or for a number of reasons need to be phased out of rotation.

File: contracts/Fund.sol
Comment (rheeger) [reply contracts/Fund.sol:133]: I think preventing an overwrite is part of what OZ wanted to see from a validation/verification standpoint. That said, updating could come into play and is arguably part of the same previous feedback.  
```
In particular, allowed organizations cant be disallowed,
grants can’t be modified nor deleted after being finalized, 
and organization claims can’t be modified.
```

Figuring `updateGrant` helps address the above.

File: contracts/Fund.sol
Comment (rheeger) [reply contracts/Fund.sol:153]: ^see previous conversation

File: contracts/OrgFactory.sol
Comment (zbronstein) [reply contracts/OrgFactory.sol]: Perhaps this is an opportunity to employ the use of the Guidestar API - they update it every day, and we can use their charity check API to confirm the NPO is still in good standing

Comment (rheeger) [reply contracts/OrgFactory.sol]: In due time, it shall be @zbronstein !! Prolly won't need to run that check before every grant, but yes, in due time we'd wire that up to update an org on chain thru this method.

File: contracts/Org.sol
Comment (mds1) [reply contracts/Org.sol]: @rheeger Your description was accurate! And you are correct that the current approach does prevent that attack

I'm good with @apbendi's path forward as long as it's good with everyone else. The additional gas costs at deployment for adding one function are relatively small compared to the total deployment costs. So as long as `rejectClaim` is guarded with the proper modifier (which seems to be the case) then there's really no additional risk of including it

File: contracts/Fund.sol
Comment (mds1) [reply contracts/Fund.sol]: Even if the total count is necessary, I think it would be be easier and cheaper to increment/decrement a `uint256`. However, we could also get the count off-chain by just fetching the event logs. I'm not sure of the use case for count so curious to hear @rheeger and @CantelopePeel's thoughts

File: contracts/OrgFactory.sol
Comment (mds1) [reply contracts/OrgFactory.sol]: It seems we are in agreement that this functionality is worthwhile, so I think we can mark this conversation resolved. Please unresolve it if you disagree!

File: contracts/Fund.sol
Comment (mds1) [reply contracts/Fund.sol:133]: I also like the separate `updateGrant` function because that explicitness shows in event logs as well. Having a `GrantCreated` event emitted when you are updating a grant would be odd. So it seems we agree this functionality is useful, so I'll mark this discussion resolved for now

File: contracts/Fund.sol
Comment (mds1) [reply contracts/Fund.sol:153]: In the words of @rheeger: ^see previous conversation 😝

(@CantelopePeel or anyone else, feel free to unresolve as always. This PR is fairly big so to help push it towards mergeable state I'm resolving discussions that don't seem contentious)

File: contracts/Org.sol
Comment (mds1) [inline contracts/Org.sol]: When rebased against the master, the rest of the inputs on this function will have validation to prevent empty strings. I suggest we rebase against the master now, then add that check on `claimId` here as well.

This has the side effect of preventing the need from checking for an empty `claimId` in `rejectClaim`

File: contracts/Org.sol
Comment (mds1) [reply contracts/Org.sol]: Here's my proposed solution: https://github.com/endaoment/endaoment-contracts/pull/75/files#r466737032

If we verify the `claimId` is not empty when it's first created, that way you have no need to check it here

File: contracts/Org.sol
Comment (mds1) [reply contracts/Org.sol:104]: I think we can resolve this discussion, or are there any outstanding issues here?

File: contracts/Fund.sol
Comment (wildmolasses) [reply contracts/Fund.sol]: +1... think we can probably get rid of count altogether (as Ben did for claims in Org.sol)

Comment (CantelopePeel) [reply contracts/Fund.sol]: Agreed, we do not need an array or even a count to maintain length... we will grab that using events.

File: contracts/Org.sol
Comment (CantelopePeel) [reply contracts/Org.sol]: I approve of @apbendi's deal. Let's keep it the way it is with the mapping.

File: contracts/Org.sol
Comment (CantelopePeel) [reply contracts/Org.sol]: Sounds good.

File: contracts/Fund.sol
Comment (rheeger) [reply contracts/Fund.sol]: We were using the count in `-client` to increment & store the index position of a given grant. If we can get a `Grant` by UUID now instead, that removes the need for the counter.

File: contracts/Org.sol
Comment (rheeger) [reply contracts/Org.sol]: Did this rebase and update happen @mds1 ? Just looking to close out convos

File: contracts/Fund.sol
Comment (mds1) [reply contracts/Fund.sol]: Awesome, Ed and I will remove that today and resolve this discussion afterwards

File: contracts/Org.sol
Comment (mds1) [reply contracts/Org.sol]: It will be happening today!

File: contracts/Org.sol
Comment (mds1) [reply contracts/Org.sol]: We ended up checking that `claimId` is not empty in both the create step and reject step

Comment (mds1) [reply contracts/Org.sol]: Left this as is 👍

**Comment by mds1** [comment]
Ok this is ready for review! Some notes:

- `FundFactory` still tracks all deployments using an array called `createdFunds`. `OrgFactory` does not do the equivalent. I think we can remove this from `FundFactory since we can track using events off-chain but want to confirm
- We moved from `memory` to `calldata` for function inputs since we don't modify them and this should save gas

**Review by CantelopePeel — APPROVED** [review]
Several more changes were made. We removed the array `createdFunds` in FundFactory and added a new modifier `onlyAddressOrAdminOrRole` to supplant the `restricted` modifier. The latter is to allow the admin to manipulate grants on behalf of a fund. This is in the event the manager abandons the fund or for other administrative tasks.

**Pull request merged**
